src/Cake/Infrastructure/Scripting/ScriptAssemblyResolver.cs

```diff
@@ -28,46 +29,74 @@ public void Dispose()
 
         private Assembly AssemblyResolve(object sender, ResolveEventArgs args)
         {
-            var name = new AssemblyName(args.Name);
-
-            // Prevent recursion from the Assembly.Load() call inside
-            if (_resolvedNames.Add(name.Name))
+            var fullName = args?.Name;
+            if (string.IsNullOrEmpty(fullName))
             {
-                _log.Verbose($"Resolving assembly {args.Name}");
+                return null;
+            }
+
+            var assemblyName = new AssemblyName(fullName);
+            var shortName = assemblyName.Name;
+            var version = assemblyName.Version.ToString();
 
-                return AssemblyResolve(name);
+            // Preventing indirect recursive calls via Assembly.Load()
+            if (!_resolvedNames.Add(shortName + version))
+            {
+                return null;
             }
-            return null;
+
+            _log.Verbose($"Resolving assembly '{fullName}' using runtime installed at '{RuntimeEnvironment.GetRuntimeDirectory()}'...");
+            return AssemblyResolve(assemblyName);
         }
 
-        private Assembly AssemblyResolve(AssemblyName name)
+        private Assembly AssemblyResolve(AssemblyName assemblyName)
         {
+            var fullName = assemblyName.FullName;
+            var shortName = assemblyName.Name;
+            var version = assemblyName.Version.ToString();
```

Comment:
Why the ToString() here? Version has a comparer 

So `x.GetName().Version == version` should work?

Reply:
Yes, `.ToString()` is unnecessary here.